objectiveC-day7.txt

```diff
@@ -0,0 +1,195 @@
+ [c1 setName:@"bb"];
+ [c2 setName:@"bb"];
+
+Now consider the below assignment using manual memory management.
+c2=c1
+
+the effect is to just copy the reference of object f1 into f2. The Student object
+referenced by c2 would be lost as a result, as its value is overwritten. This creates a
+ memory leak  (a variable that is no longer referenced and therefore can’t be released).
+If you’re using ARC, c1 and c2 are both strong variables. So, the previous assignment actually
+works like this:
+ [c1 retain]; // retain new value
+ [c2 release]; // release the old value
+ c2 = c1; // copy the reference
+ --The above steps are done by the compiler automatically.
+-Properties that we write inside a class are not Strong by default.We have to declare them explicitly.
+example: @property(strong) NSMutableArray *employeList
+
+Weak variables:
+-Weak variables come into usage when there is a retain cycle.Retain cycle is created when two objects have reference to each other.
+When a parent class holds a reference to a subclass and subclass holds a reference to parent class,a cycle is created and these objects can
+never be destroyed.So,we need to make another type of reference known as weak reference between two objects.
+--A weak variable does not prevent deallocation of the object it references.
+Syntac: @property (weak) NSMutableArray *employeList;
+
+--weak variables are not supported in iOS 4 or Mac OS X 10.6. In such cases, you can
+still use the unsafe_unretained (or assign) property attribute or declare your variable to be __unsafe_unretained.
+
+------------------------------------------------------------------------------------------------------------------------------------------------------------
+
+3)
+When we declare a property 'weak' ,it does not prevent deallocation of the object it references in case of retain cycle.
+
+When you declare a weak variable , the system tracks the reference that is made on assignment to that variable,and when that referenced object gets deallocated,
+the weak variable gets automatically set to nil. That prevents any crashes that might occur by  sending a message to that variable. Because the variable
+will be set to nil ,sending a message to a nil object does nothing, thus preventing a crash.
+
+To declare a weak variable, you use the __weak keyword:
+__weak NSString *name;
+or you use the weak attribute for a property:
+ @property (weak) NSString *name;
+ 
+weak variables are not supported in iOS 4 or Mac OS X 10.6. In such cases, you can
+still use the unsafe_unretained (or assign) property attribute or declare your variable to be
+__unsafe_unretained.
+--Note that these variables are not zeroed automatically when the referenced object is deallocated. So,there is danger of crashing the app,when messages
+is passed to a object which is deallocated.
+--------------------------------------------------------------------------------------------------------------------------------------------------------
```

Comment:
When we se property as weak, we cannot guarantee that its value is nonnull by default. If it is set to nil, it may cause miscalculation or may cause app crash.